---
title: 'CA2200: следует повторно вызывать исключение для сохранения сведений о стеке (анализ кода)'
description: 'Сведения о правиле анализа кода "CA2200: следует повторно вызывать исключение для сохранения сведений о стеке"'
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- RethrowToPreserveStackDetails
- CA2200
helpviewer_keywords:
- CA2200
- RethrowToPreserveStackDetails
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
ms.openlocfilehash: c49f3dbab810ce32bfab2d13fbbe9c9f79da6df5
ms.sourcegitcommit: 05d0087dfca85aac9ca2960f86c5efd218bf833f
ms.translationtype: HT
ms.contentlocale: ru-RU
ms.lasthandoff: 03/30/2021
ms.locfileid: "99719933"
---
# <a name="ca2200-rethrow-to-preserve-stack-details"></a><span data-ttu-id="0aff0-103">CA2200. Повторно порождайте исключения для сохранения сведений стека</span><span class="sxs-lookup"><span data-stu-id="0aff0-103">CA2200: Rethrow to preserve stack details</span></span>

| | <span data-ttu-id="0aff0-104">Значение</span><span class="sxs-lookup"><span data-stu-id="0aff0-104">Value</span></span> |
|-|-|
| <span data-ttu-id="0aff0-105">**Идентификатор правила**</span><span class="sxs-lookup"><span data-stu-id="0aff0-105">**Rule ID**</span></span> |<span data-ttu-id="0aff0-106">CA2200</span><span class="sxs-lookup"><span data-stu-id="0aff0-106">CA2200</span></span>|
| <span data-ttu-id="0aff0-107">**Категория**</span><span class="sxs-lookup"><span data-stu-id="0aff0-107">**Category**</span></span> |[<span data-ttu-id="0aff0-108">Использование</span><span class="sxs-lookup"><span data-stu-id="0aff0-108">Usage</span></span>](usage-warnings.md)|
| <span data-ttu-id="0aff0-109">**Исправление является критическим или не критическим**</span><span class="sxs-lookup"><span data-stu-id="0aff0-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="0aff0-110">Не критическое</span><span class="sxs-lookup"><span data-stu-id="0aff0-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="0aff0-111">Причина</span><span class="sxs-lookup"><span data-stu-id="0aff0-111">Cause</span></span>

<span data-ttu-id="0aff0-112">В операторе `throw` повторно вызывается и явным образом задается исключение.</span><span class="sxs-lookup"><span data-stu-id="0aff0-112">An exception is rethrown and the exception is explicitly specified in the `throw` statement.</span></span>

## <a name="rule-description"></a><span data-ttu-id="0aff0-113">Описание правила</span><span class="sxs-lookup"><span data-stu-id="0aff0-113">Rule description</span></span>

<span data-ttu-id="0aff0-114">При возникновении исключения часть информации в нем представляет собой трассировку стека.</span><span class="sxs-lookup"><span data-stu-id="0aff0-114">Once an exception is thrown, part of the information it carries is the stack trace.</span></span> <span data-ttu-id="0aff0-115">Трассировка стека — это список иерархии вызовов методов, который начинается с метода, вызывающего исключение, и завершается методом, перехватывающим исключение.</span><span class="sxs-lookup"><span data-stu-id="0aff0-115">The stack trace is a list of the method call hierarchy that starts with the method that throws the exception and ends with the method that catches the exception.</span></span> <span data-ttu-id="0aff0-116">Если исключение повторно вызывается путем указания этого исключения в операторе `throw`, в текущем методе трассировка стека перезапускается, а список вызовов метода между исходным методом, создавшим исключение, и текущим методом теряется.</span><span class="sxs-lookup"><span data-stu-id="0aff0-116">If an exception is re-thrown by specifying the exception in the `throw` statement, the stack trace is restarted at the current method and the list of method calls between the original method that threw the exception and the current method is lost.</span></span> <span data-ttu-id="0aff0-117">Для сохранения исходных данных трассировки стека с исключением следует использовать оператор `throw` без указания исключения.</span><span class="sxs-lookup"><span data-stu-id="0aff0-117">To keep the original stack trace information with the exception, use the `throw` statement without specifying the exception.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="0aff0-118">Устранение нарушений</span><span class="sxs-lookup"><span data-stu-id="0aff0-118">How to fix violations</span></span>

<span data-ttu-id="0aff0-119">Чтобы устранить нарушение этого правила, повторно вызывайте это исключение без его явного указания.</span><span class="sxs-lookup"><span data-stu-id="0aff0-119">To fix a violation of this rule, rethrow the exception without specifying the exception explicitly.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="0aff0-120">Условия для отключения предупреждений</span><span class="sxs-lookup"><span data-stu-id="0aff0-120">When to suppress warnings</span></span>

<span data-ttu-id="0aff0-121">Для этого правила отключать вывод предупреждений не следует.</span><span class="sxs-lookup"><span data-stu-id="0aff0-121">Do not suppress a warning from this rule.</span></span>

## <a name="example"></a><span data-ttu-id="0aff0-122">Пример</span><span class="sxs-lookup"><span data-stu-id="0aff0-122">Example</span></span>

<span data-ttu-id="0aff0-123">В следующем примере показан метод`CatchAndRethrowExplicitly`, нарушающий правило, и метод `CatchAndRethrowImplicitly`, соответствующий правилу.</span><span class="sxs-lookup"><span data-stu-id="0aff0-123">The following example shows a method, `CatchAndRethrowExplicitly`, which violates the rule and a method, `CatchAndRethrowImplicitly`, which satisfies the rule.</span></span>

:::code language="csharp" source="snippets/csharp/all-rules/ca2200.cs" id="snippet1":::

:::code language="vb" source="snippets/vb/all-rules/ca2200-rethrow-to-preserve-stack-details_1.vb":::
